### PR TITLE
Add an integration test for one of the happy-paths for release-log

### PR DESCRIPTION
**Pull request by Limess**
There weren't any tests for this handler.

Added tests using nock to stub all API calls, and calling the handler with a mock kinesis record.

This is relatively brittle due to the nature of the calls, and also involves stubbing the schema using a 'private'ish API as the library doesn't expose the ability to use stub rawData - this should be a thing.

I've used the prometheus-pingdom-exporter runbook as a fixture for now as it's relatively flushed out but this should probably be bespoke in future.

Additionally moved the babel config to be shared between webpack server/jest as jest blew up when getting near JSX.

~N.B this is branched from https://github.com/Financial-Times/runbook.md/pull/97 as it tests currently invalid repo names.~

### CODE REVIEW

File: babel.config.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+'use strict';
+
+// server/test config
+module.exports = {
```
Comment (wheresrhys) [inline babel.config.js]: Why did we get away without this before/why is it necessary now?

File: lambdas/ingester/src/lib/external-apis.js [hunk 2]
```diff
@@ -17,7 +20,10 @@ const callExternalApi = async ({
 		body: JSON.stringify(payload),
 		headers,
 	};
-	const fetchResponse = await nodeFetch(url, options);
+	const fetchResponse = await nodeFetch(url, {
+		...options,
+		agent: keepAliveAgent,
```
Comment (wheresrhys) [inline lambdas/ingester/src/lib/external-apis.js]: An optimisation, or does this fix something in particular? Are there any risks to using keep alive in a lambda?

File: lambdas/ingester/src/releaseLogHandler.js [hunk 3]
```diff
@@ -217,7 +221,7 @@ const fetchRunbookMds = (parsedRecords, childLogger) =>
 				const [gitRepositoryName] =
 					gitRefUrl
 						.replace('https://github.com/', '')
-						.match(/[\w-]+\/[\w-]+/) || [];
+						.match(/[a-z0-9_.-]+\/[a-z0-9_.-]+/i) || [];
```
Comment (wheresrhys) [inline lambdas/ingester/src/releaseLogHandler.js]: Last time I checked, github had no published restrictions on characters in repo names. Arguably this regex should use `[^/]*`, but (unless you can find a reference for repo naming limitations), shouldn't be made struicter.

Or is this a git limitation, rather than github?

File: lambdas/ingester/src/releaseLogHandler.js [hunk 4]
```diff
@@ -264,14 +268,22 @@ const fetchRunbookMds = (parsedRecords, childLogger) =>
 					},
 					'Retrieving runbook.md from GithubApi has failed',
 				);
+				throw Object.assign(
```
Comment (wheresrhys) [inline lambdas/ingester/src/releaseLogHandler.js]: this new error, I take it, is the meat of this PR?

File: lambdas/ingester/src/releaseLogHandler.js [hunk 5]
```diff
@@ -304,7 +316,19 @@ const handler = async (event, context) => {
...
 
-	await processRunbookMd(parsedRecords, childLogger);
+	if (parsedRecords.length === 0) {
+		childLogger(
+			{
+				event: 'SKIPPING_INGEST',
+			},
+			'Nothing to ingest, skipping event',
+		);
+		return json(200, {
```
Comment (wheresrhys) [inline lambdas/ingester/src/releaseLogHandler.js]: along with this

File: lambdas/ingester/test/fixtures/runbook-fixture.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
+<!--
+    Written in the format prescribed by https://github.com/Financial-Times/runbook.md.
```
Comment (wheresrhys) [inline lambdas/ingester/test/fixtures/runbook-fixture.md]: this is actually not a bad thing to encourage people to do in general. We could add it to our example content

File: lambdas/ingester/test/releaseLogHandler.test.js [hunk 7]
```diff
@@ -0,0 +1,261 @@
+const nock = require('nock');
```
Comment (wheresrhys) [inline lambdas/ingester/test/releaseLogHandler.test.js]: quibbling the naming of this test file - looks like there's a lot of github app type stuff in here

File: secret-squirrel.js [hunk 8]
```diff
@@ -1,16 +1,16 @@
...
+		allow: [],
 		allowOverrides: []
 	},
 	strings: {
 		deny: [],
 		denyOverrides: [
-			'b62f4580-81f6-11e9-8f9b-7d694d159e85',
 			'b62f4580-81f6-11e9-8f9b-7d694d159e85', // README.md:9
-			'baxterthehacker@users.noreply.github.com'
+			'baxterthehacker@users\\.noreply\\.github\\.com', // githubPushEvent.json:20|25|42|47|60|130, lambdas/ingester/test/fixtures/githubPushEvent.json:20|25|42|47|60|130
```
Comment (wheresrhys) [inline secret-squirrel.js]: what's this guy's backstory? What's his motivation?

File: babel.config.js (hunk 1 above)
Comment (Limess) [reply babel.config.js]: Jest was never loading the client side files with JSX in them. As soon as it hits a code path that tries to load them (this handler for example) it blows up as it can't read the syntax.

Adding `babel-jest` and a config so it can read the files resolves this.

File: lambdas/ingester/src/lib/external-apis.js (hunk 2 above)
Comment (Limess) [reply lambdas/ingester/src/lib/external-apis.js]: Optimisation, probably premature.

No risks that I know of, if the socket is closed it'll just open a new one.

File: lambdas/ingester/src/releaseLogHandler.js (hunk 3 above)
Comment (Limess) [reply lambdas/ingester/src/releaseLogHandler.js]: See https://github.com/Financial-Times/runbook.md/pull/97, there's a link to a moby (Docker) PR where they use this pattern.

I believe it's that github convert any other characters to these characters in URLs, if you have evidence otherwise then yes, `[^\/]` makes sense (I was going to use that prior to googling it).

File: lambdas/ingester/src/releaseLogHandler.js (hunk 4 above)
Comment (Limess) [reply lambdas/ingester/src/releaseLogHandler.js]: One of the concrete changes yes. I wouldn't call it the meat though...

Just what I found while writing tests and trying to make them pass or give reasonable errors.

File: lambdas/ingester/src/lib/external-apis.js (hunk 2 above)
Comment (wheresrhys) [reply lambdas/ingester/src/lib/external-apis.js]: My worry is this might mean we get lots of weird, confusing-when-debugging socket closure error logs (we did in runbooks when doing some not dissimilar out of band requesting)

File: lambdas/ingester/src/releaseLogHandler.js (hunk 5 above)
Comment (Limess) [reply lambdas/ingester/src/releaseLogHandler.js]: This doesn't actually do anything as nothing is inspecting the lambda output, but it does mean it's testable.

File: lambdas/ingester/test/fixtures/runbook-fixture.md (hunk 6 above)
Comment (Limess) [reply lambdas/ingester/test/fixtures/runbook-fixture.md]: It's relatively common to have a header like this for machine generated files (puppet/ansible etc), I think it's a good idea to add.

File: lambdas/ingester/test/releaseLogHandler.test.js (hunk 7 above)
Comment (Limess) [reply lambdas/ingester/test/releaseLogHandler.test.js]: It's matching the name of the handler file it's testing. There are a lot of naming issues in this repo (things not named as they are doing, inconsistent naming conventions in the same folder).

File: lambdas/ingester/src/lib/external-apis.js (hunk 2 above)
Comment (Limess) [reply lambdas/ingester/src/lib/external-apis.js]: I haven't seen this anywhere else - wasn't that down to making head requests or not reading the body on gets?

File: lambdas/ingester/src/releaseLogHandler.js (hunk 3 above)
Comment (wheresrhys) [reply lambdas/ingester/src/releaseLogHandler.js]: yep they appear to. Frustrating it's not documented anywhere

File: lambdas/ingester/test/releaseLogHandler.test.js (hunk 7 above)
Comment (wheresrhys) [reply lambdas/ingester/test/releaseLogHandler.test.js]: Ah in that case I suppose I'm surprised that the change log handler needs to interact with sos. or with pull requests. mind you, I haven't worked with the code much

File: lambdas/ingester/src/lib/external-apis.js (hunk 2 above)
Comment (wheresrhys) [reply lambdas/ingester/src/lib/external-apis.js]: yep, but the underlying problem was that sockets were not being closed, the lambda execution paused, and then the socket timed out. Same thing could happen here, no?

File: lambdas/ingester/test/releaseLogHandler.test.js (hunk 7 above)
Comment (Limess) [reply lambdas/ingester/test/releaseLogHandler.test.js]: I was very surprised by how much this thing does in a single flow, I found it very hard to follow - one of the reasons I decided to write an end-to-end integration test for the handler.

There are actually more code paths which it can hit with github if it's not a PR which this test does not cover.

File: lambdas/ingester/src/lib/external-apis.js (hunk 2 above)
Comment (Limess) [reply lambdas/ingester/src/lib/external-apis.js]: I don't believe so as nothing happens in the background, it's all in the runtime of a single lambda.

If the lambda freezes and restarts after a set of records is processed I'm fairly sure that it'll get a new socket?

**Comment by Limess** [comment]
Why a tentative 👍? Only related to adding keepAlive?
